File: offer/24.go
/*
 * @Author: lmio
 * @Date: 2023-03-23 19:52:22
 * @LastEditTime: 2023-03-23 20:30:10
 * @FilePath: /leetcode/offer/24.go
 * @Description:剑指 Offer 24. 反转链表
 */
package offer

import (
	. "leetcode/code_struct"
)

func ReverseList(head *ListNode) *ListNode {
	var prev *ListNode
	node := head
	for node != nil {
		temp := node.Next
		node.Next = prev
		prev = node
		node = temp
	}
	return prev
}

File: code/24.py
'''
Author: lmio 2091319361@qq.com
Date: 2023-08-06 20:19:33
LastEditors: lmio 2091319361@qq.com
Description: 24. 两两交换链表中的节点
'''

from typing import Optional

from utils.node import ListNode


class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(next=head)
        fist = dummy
        if not head:
            return None
        pre, cur = head, head.next
        while cur:
            pre.next = cur.next
            cur.next = pre
            fist.next = cur
            fist = pre
            pre = pre.next
            if not pre:
                break
            cur = pre.next
        return dummy.next